apps/blockchain/test/blockchain_test.exs

```diff
@@ -27,22 +27,86 @@ defmodule BlockchainTest do
   }
 
   test "runs blockchain tests" do
-    Enum.each(tests(), fn json_test_path ->
-      json_test_path
-      |> read_test()
-      |> ignore_known_failing_tests(json_test_path)
-      |> Enum.each(&run_test/1)
+    forks_with_existing_implementation()
+    |> Enum.map(&spawn_forks/1)
+    |> Enum.flat_map(&receive_replies/1)
+    |> make_assertions()
+  end
+
+  defp spawn_forks(fork) do
+    {fork_pid, fork_ref} = spawn_tests_for_fork(fork)
+    {fork, fork_pid, fork_ref}
+  end
+
+  defp receive_replies({fork, fork_pid, fork_ref}) do
+    ten_minute_timeout = 1000 * 60 * 10
+
+    case receive_fork_reply(fork_pid, fork_ref, ten_minute_timeout) do
+      {:fork_failure, error} ->
+        raise fork_failure_error(fork, error)
+
+      {:fork_timeout, stacktrace} ->
+        raise fork_timeout_error(fork, stacktrace, ten_minute_timeout)
+
+      test_failures ->
+        test_failures
+    end
+  end
+
+  defp fork_failure_error(fork, error) do
+    "[#{fork}] error: #{inspect(error)}"
+  end
+
+  defp fork_timeout_error(fork, stacktrace, timeout) do
+    "[#{fork}] timeout after #{inspect(timeout)} milliseconds: #{inspect(stacktrace)}"
+  end
+
+  defp spawn_tests_for_fork(fork) do
+    parent = self()
+
+    spawn_monitor(fn ->
+      failing_tests = run_tests_for_fork(fork)
+      send(parent, {self(), :fork_tests_finished, failing_tests})
+      exit(:shutdown)
     end)
   end
 
-  defp ignore_known_failing_tests(tests, json_test_path) do
-    Enum.filter(tests, fn {_name, test} ->
-      !known_failing_test?(json_test_path, test)
+  defp receive_fork_reply(fork_pid, fork_ref, timeout) do
+    receive do
+      {^fork_pid, :fork_tests_finished, failing_tests} ->
+        Process.demonitor(fork_ref, [:flush])
+        failing_tests
+
+      {:DOWN, ^fork_ref, :process, ^fork_pid, error} ->
+        {:fork_failure, error}
+    after
+      timeout ->
+        case Process.info(fork_pid, :current_stacktrace) do
+          {:current_stacktrace, stacktrace} ->
+            Process.demonitor(fork_ref, [:flush])
+            Process.exit(fork_pid, :kill)
+            {:fork_timeout, stacktrace}
+
+          nil ->
+            receive_fork_reply(fork_pid, fork_ref, timeout)
+        end
+    end
+  end
+
+  defp run_tests_for_fork(fork) do
+    tests()
+    |> Stream.reject(&known_fork_failures?(&1, fork))
+    |> Enum.flat_map(fn json_test_path ->
+      json_test_path
+      |> read_test()
+      |> Stream.filter(&fork_test?(&1, fork))
```

Comment:
what is justification for using `Stream` instead of `Enum`?

Reply:
Since we're creating a lot of intermediate lists in each step, and the list of tests can be big, I thought we could benefit from lazily evaluating those things. That way we aren't constantly building lists over and over again. I'm happy to bring it back to `Enum` if it makes more sense, but it seemed like a good use case to me.